### ISSUE TITLE
xmlObject on referenced type sets name of property

### ISSUE THREAD

**Issue by emilw** [NJsonSchema/NJsonSchema#299]
Hi,
There is a bug that appears when a property referencing a type with an xml object set on it, then the xml name of the property get's the same name as the type.
(This behaviour is discussed here: https://github.com/swagger-api/swagger-ui/issues/2610)

**Example and how to reproduce it:**
Here is a property defined on class X, it's called AccrualCodeString and references IntegrationCodingDimensionDto.
```c#
public IntegrationCodingDimensionsDto AccrualCodeString { get; set; }
.
.
.
```
Here is the IntegrationCodingDimensionDto defined, with an XmlType attribute to control how the root element should be named when that class is used on it's own:
```c#
[XmlType("CodingDimensions")]
    public class IntegrationCodingDimensionsDto
    {
.
.
.
```
When serializing object of type X to an XML it will look as bellow:
```xml
<X>
<AccrualCodeString >
    <Dimension1>string</Dimension1>
    <Dimension2>string</Dimension2>
 .
 .
  </AccrualCodeString >
</X>
```
Current swagger ouput
```json
"AccrualCodeString": {
                    "$ref": "#/definitions/IntegrationCodingDimensionsDto",
}
```
When retrieving example model for XML in Swagger UI it gives me:
```xml
<X>
<CodingDimensions>
    <Dimension1>string</Dimension1>
    <Dimension2>string</Dimension2>
.
.
.
  </CodingDimensions>
</X>
```
So it more or less got the type name as property name.

We need to generate this swagger property definition to make it work:
```json
"AccrualCodeString": {
"$ref": "#/definitions/IntegrationCodingDimensionsDto",
"xml":{
"name":"AccrualCodeString"
 }
},
```
Then i will get the right output:
```xml
<AccrualCodeString >
    <Dimension1>string</Dimension1>
    <Dimension2>string</Dimension2>
.
.
  </AccrualCodeString >
```
The question is if we should play it safe and set the xml property every time for a reference type property or check if the type that it references have a XMLType attribute set?
I can create a PR and add the corresponding test and the new behaviour.

**Comment by rsuter** [comment]
I'd go for the latter, check if the xml attribute exists. I think most users (me too) have JSON only services and would prefer to have a Swagger spec without XML specs in it...

**Comment by rsuter** [comment]
I agree with you: Its an odd behavior... I'd implement it this way: Generate the xml property name if the referenced type or schema overrides the name and would cause xml with the wrong tag

**Comment by rsuter** [comment]
Just checked your proposal and this is not allowed: 

```
"AccrualCodeString": {
"$ref": "#/definitions/IntegrationCodingDimensionsDto",
"xml":{
"name":"AccrualCodeString"
 }
},
```

You can either have $ref or other properties, but not both... (either you reference something or you define it locally, you cannot mix these two...)

this is why i think Swagger UI does it wrong... or at least, this is once again a fundamental problem of the Swagger spec.

**Comment by emilw** [comment]
Hi,
Good that you saw it the same, i was 100% pure that it was an issue in Swagger UI, that was why i registered the bug here: https://github.com/swagger-api/swagger-ui/issues/2610

But i got it confirmed that it was suppose to work that way, so i guess we need to align :-).

About the solution, I tested with the approach above and it works(see the source line):
https://github.com/emilw/MixedStuff/blob/master/Development/Web/OpenAPI/ControllersSwagger.json#L33

Tested it like this in Swagger UI:
http://petstore.swagger.io
And swagger url:
https://raw.githubusercontent.com/emilw/MixedStuff/master/Development/Web/OpenAPI/ControllersSwagger.json
Check Post method of AccrualTemplateImportBatch, select application/xml to produce.

Where do you see that it's not allowed? "$ref" is a sub property to the property, so it should be doable right? It looks like it merge the pulled in property with the one already there from my point of view...

**Comment by rsuter** [comment]
Can we close this?

**Comment by emilw** [comment]
Yes

**Comment by rsuter** [comment]
I think instead of 

```
"AccrualCodeString": {
	"$ref": "#/definitions/IntegrationCodingDimensionsDto",
	"xml":{
		"name":"AccrualCodeString"
	}
}
```
We need to use: 

```
"AccrualCodeString": {
	oneOf: [
		{
			"$ref": "#/definitions/IntegrationCodingDimensionsDto",
		}
	],
	"xml":{
		"name":"AccrualCodeString"
	}
}
```

This would be a correct... Can you test if this is supported by Swagger UI?

**Comment by emilw** [comment]
Hi,
Sorry for late reply, i will try to plan it this week. I have had full time keeping up with my kids etc. I get back to you when i have tested it.

**Comment by rsuter** [comment]
Ok no problem. I think ill release njs and nswag today, but we can fix this later...

**Comment by emilw** [comment]
Hi,
I have tested now:
With the petstore swagger UI it do not work.
I made the change according to your suggestion, see change on line here:
https://github.com/emilw/MixedStuff/blob/master/Development/Web/OpenAPI/ControllersSwaggerWithOneOf.json#L35
I executed it in the petstore Swagger UI with this one.
https://raw.githubusercontent.com/emilw/MixedStuff/master/Development/Web/OpenAPI/ControllersSwaggerWithOneOf.json

The example xml looks like this(it writes invalid xml in the message as shown below):
```xml
<?xml version="1.0"?>
<AccrualTemplate>
  <Name>string</Name>
  <CompanyId>string</CompanyId>
  <!-- invalid XML -->
  <AccrualPeriods>
    <AccrualPeriods>
.
.
```

If i run the same test with the one we have in right now i get no problem. That one is here:
https://raw.githubusercontent.com/emilw/MixedStuff/master/Development/Web/OpenAPI/ControllersSwagger.json

Can you check so i did not misunderstand anything? To be honest i still don't see why this is not ok, is it because we rely on incorrect logic in Swagger UI or?

**Comment by rsuter** [comment]
According to https://tools.ietf.org/id/draft-pbryan-zyp-json-ref-03.html ($ref specification): 

> Any members other than "$ref" in a JSON Reference object SHALL be ignored.

**Comment by rsuter** [comment]
So yes, I don't want to rely on incorrect logic in Swagger UI...

**Comment by emilw** [comment]
Ok, so now i get it. It states pretty clear that this behaviour is not supported.
But how do you see it with oneOf, i do not see that this makes it better? Or what is the difference there? It's still a reference that need to be pulled in..

**Comment by rsuter** [comment]
Yes, but the the schema in "allOf" only has a $ref keyword, no other properties...

**Comment by emilw** [comment]
I see that you pointed out allof, i tested with oneof 😩. I Will redo the test again...

**Comment by rsuter** [comment]
I think oneOf is not supported by Swagger, you need to use allOf

**Comment by emilw** [comment]
I somehow mixed them when testing it, my intention was to test "allOf" but somehow it became "oneOf"...let me test again and get back to you.

**Comment by emilw** [comment]
Great success!

See swagger file, and line:
https://github.com/emilw/MixedStuff/blob/master/Development/Web/OpenAPI/ControllersSwaggerWithAllOf.json#L35

Tested the raw format from this link:
https://raw.githubusercontent.com/emilw/MixedStuff/master/Development/Web/OpenAPI/ControllersSwaggerWithAllOf.json
in Swagger UI.

Sorry for messing things up earlier, i am a little bit tired :-).

I still do not understand why allOf array with $ref is more ok than a regular property $ref when using the xml object? Why is it ok to add:
```json
"xml":{
"name":"AccrualCodeString"
 }
```
to the property for the allOf?

/Emil

**Comment by rsuter** [comment]
The JSON format specification says that if you have a $ref property, all other properties of the object are ignored. This is why an object with a $ref and "xml" property are not allowed... Using allOf is in line with this rule.

**Comment by emilw** [comment]
So you mean that due to that the $ref property is one level beneath in the allOf case, that is why it's ok or?

Do you have any specification for the allOf, e.g. similar to the one you added above for JSON Reference(https://tools.ietf.org/id/draft-pbryan-zyp-json-ref-03.html)?

I really lack he big picture on the JSON schema structure, i really need to do some reading there. Any good material that you can recommend?

**Comment by rsuter** [comment]
JSON Schema: 
http://json-schema.org/latest/json-schema-validation.html#rfc.section.5.22

Swagger spec: 
http://swagger.io/specification/#composition-and-inheritance--polymorphism--83

**Comment by WillCallahan** [comment]
This needs to be extended so that properties in an object with the _XmlAttributeAttribute_ are given the `"xml": { "attribute": true }` field.

> Example

```csharp
[XmlType("Pet")]
public class Pet {
    [XmlAttribute("name")]
    public string Name { get; set; }

    [XmlAttribute("name")]
    public Address Address { get; set; }
}
```

Should produce

```json
{
  "Pet": {
    "type": "object",
    "properties": {
      "name": {
        "type": "string"
        "xml": {
          "name": "name",
          "attribute": true
        }
      },
      "address": {
        "$ref": "#/definitions/Address"
      }
    }
  }
}
```

**Comment by emilw** [comment]
@WillCallahan, can you provide the example XML you are after just to get it from end to end?
We are also lacking the ability to handle a collection/List/Array parameter for XML, do you have any input on where we need to put the xml object in that case? Is it on the parameter or on the data definition? Right now it just add the element and not the surrounded e.g. "ArrayOfItems" or similar.

**Comment by emilw** [comment]
@RSuter , this is working as the original ticket was created i dont see why it's still open?
The ArrayOfItems issue mentioned above is registered here: https://github.com/RSuter/NJsonSchema/issues/777
